### PR TITLE
Issue 199

### LINKED ISSUE #199: Assertion on existing keys
Ideally the double sided map should have two insert methods, once that asserts if the key is present and one that overrides. This way we can ensure we do the correct thing. The same goes for insertion into unique maps (see localvar7 PR)

### PR DESCRIPTION
**Pull request by mroethlin**
## Technical Description
 
Current proposal for issue 199. DoubleSidedMap now features two insertion methods, one which asserts that elements are not already present, and another one which will silently reject the insertion if the element is already present. Appropriate methods are called from StencilMetaInformation. Furthermore, all emplace calls to unordered_maps in StencilMetaInformation were prepended with an assert that the element is not already present"

 #### Resolves
 
#199 
 
 ## Dependencies
 
 This PR is independent.

### CODE REVIEW

File: src/dawn/CodeGen/CXXNaive/ASTStencilBody.cpp [hunk 1]
```diff
@@ -94,7 +94,9 @@ void ASTStencilBody::visit(const std::shared_ptr<UnaryOperator>& expr) { Base::v
 
 void ASTStencilBody::visit(const std::shared_ptr<BinaryOperator>& expr) { Base::visit(expr); }
 
-void ASTStencilBody::visit(const std::shared_ptr<AssignmentExpr>& expr) { Base::visit(expr); }
+void ASTStencilBody::visit(const std::shared_ptr<AssignmentExpr>& expr) {
```
Comment (twicki) [inline src/dawn/CodeGen/CXXNaive/ASTStencilBody.cpp]: seems like your formatting is either incompatible or you're using something different?

File: src/dawn/IIR/StencilMetaInformation.cpp [hunk 2]
```diff
@@ -579,8 +582,8 @@ int StencilMetaInformation::getStencilIDFromStencilCallStmt(
 }
 
 void StencilMetaInformation::insertStencilCallStmt(std::shared_ptr<StencilCallDeclStmt> stmt,
-                                                   int stencilID) {
-  StencilIDToStencilCallMap_.emplace(stencilID, stmt);
+                                                   int stencilID) {
```
Comment (twicki) [inline src/dawn/IIR/StencilMetaInformation.cpp]: format?

File: src/dawn/Support/DoubleSidedMap.h [hunk 3]
```diff
@@ -26,10 +26,25 @@ class DoubleSidedMap {
...
+  /// present. That is, one to one relation (bijective mapping) is maintained
+  void add(const Key1& key1, const Key2& key2) {
+    DAWN_ASSERT(!directHas(key1));
+    DAWN_ASSERT(!reverseHas(key2));
+
+    directMap_.emplace(key1, key2);
+    reverseMap_.emplace(key2, key1);
+  }
+
+  /// Insert a new element into the map. This function works analogous to the emplace method in
```
Comment (havogt) [inline src/dawn/Support/DoubleSidedMap.h]: emplace has to be deleted as a non-bijective map breaks the erase functions.

File: src/dawn/Support/DoubleSidedMap.h [hunk 4]
```diff
@@ -26,10 +26,25 @@ class DoubleSidedMap {
   std::unordered_map<Key2, Key1> reverseMap_;
 
 public:
+
+  /// Add a new element to the map. This function asserts that neither key1 nor key2 is already
+  /// present. That is, one to one relation (bijective mapping) is maintained
+  void add(const Key1& key1, const Key2& key2) {
+    DAWN_ASSERT(!directHas(key1));
```
Comment (havogt) [inline src/dawn/Support/DoubleSidedMap.h]: This should not be an assert. It should be checked also in release...

**Comment by Stagno** [comment]
will this PR fix this issue? https://github.com/MeteoSwiss-APN/dawn/issues/210

**Comment by mroethlin** [comment]
launch jenkins

**Pull request merged**